In [45]:
import numpy as np
from scipy import stats

# Question 1

In [6]:
file = open("CS_412_Fall21_Assignment1/data.online.scores.txt")
data = np.loadtxt(file, unpack='true')
file.close()

In [7]:
data

array([[   0.,    1.,    2., ...,  997.,  998.,  999.],
       [  97.,   83.,   79., ...,   85.,   84.,   56.],
       [  95.,   99.,   81., ...,   91.,   99.,   87.]])

In [49]:
mid = data[1]

## (a) maxmum is 100, minimum is 37

Use max and min function to find the max number and min number in 1-D array mid

In [31]:
mid.max()

100.0

In [19]:
mid.min()

37.0

## (b) First quartile Q1 = 68, median = 77, and third quartile Q3 = 87.

sort number from smallest to largest, and first quartile is the number in the 25% position, median number sit in the 50% position while third quartile sit in the 75% position.  

In [20]:
np.percentile(mid, 25)  # Q1

68.0

In [21]:
np.percentile(mid, 50)  # median

77.0

In [22]:
np.percentile(mid, 75)  # Q3

87.0

## (c) Mean = 76.715

sum all the numer in the array mid and then divide 100 which is the number in the array.

In [23]:
np.mean(mid)

76.715000000000003

## (d) Mode = 77, which occurs 37 times. 

Mode is the number with the most occurrence in the array mid. 

In [29]:
stats.mode(mid)

ModeResult(mode=array([ 77.]), count=array([37]))

## (e) Variance = 173.106

sum (x-mean)^2 and then divided by N which is 100 to get variance.

In [32]:
np.var(mid)

173.10577500000002

# Qestion 2

In [33]:
Area = 46+64+79+92+65+53+35+32+27
Area/2.0

246.5

Delete area in 10-25 hours and we get 57.5 < 92 which is in 25-30 hours

In [35]:
246.5-46-64-79

57.5

Therefore, we know the median is lacated within 25-30 hours.

In [36]:
(30-25)/92*57.5+25

28.125

### median hourly pay at SkyNet is about 28.125 dollars per hour

# Question 3

First calculate normalized midterm array

z-score normalization: for every element in array mid, first minus mean, then divided by standard deviation of array mid.

In [48]:
normalized_mid = (mid-76.715)/np.std(mid)
normalized_mid

array([ 1.54176873,  0.47769369,  0.17367225, -0.35836527,  1.16174193,
        0.09766689,  0.17367225, -0.66238671,  0.09766689,  0.02166153,
       -0.13034919,  1.00973121,  0.62970441,  0.02166153,  0.93372585,
        0.70570977, -0.8904028 , -0.73839207, -0.73839207,  1.23774729,
        0.47769369,  1.54176873,  0.70570977, -1.19442424,  0.62970441,
        1.46576337,  0.17367225, -1.19442424,  1.23774729,  1.61777409,
        0.55369905, -1.72646176,  0.85772049,  0.70570977,  0.02166153,
        1.08573657,  1.38975801,  0.24967761, -1.11841888, -0.96640816,
        0.24967761,  1.31375265,  1.00973121,  0.40168833, -0.20635455,
       -1.34643496, -0.43437063,  1.16174193,  0.17367225,  1.31375265,
       -1.19442424, -0.05434383,  1.31375265,  0.47769369, -2.86654216,
        0.24967761, -1.80246712,  1.76978481,  0.09766689,  1.54176873,
       -0.58638135, -1.6504564 ,  1.00973121, -0.13034919, -1.42244032,
        1.23774729,  0.32568297, -0.66238671,  0.02166153, -0.28

## (a) Variance of midterm-original = 173.106, Variance of midterm-normalized = 1

sum (x-mean)^2 and then divided by N which is 100 to get variance.

In [52]:
np.var(normalized_mid)

0.99999999999999978

In [53]:
np.var(mid)

173.10577500000002

## (b) Given 90, the corresponding score after normalization is 1.009731210573523

In [54]:
(90-np.mean(mid))/np.std(mid) # first minus mean then divided by standard deviation

1.009731210573523

In [55]:
normalized_mid[11] # Also, which is student id 11

1.009731210573523

## (c) Pearson's correlation coefficient between midterm-original and finals-original is 0.544424742312

In [62]:
final = data[2]

Pearson's correlation coefficient = covariance(X, Y) / (stdv(X) * stdv(Y))
    Here X is mid, Y is final

In [63]:
corr, _ = stats.pearsonr(mid, final)
print(corr)

0.544424742312


## (d) Pearson's correlation coefficient between midterm-normalized and finals-original is 0.544424742312

Pearson's correlation coefficient = covariance(X, Y) / (stdv(X) * stdv(Y))
Here X is normalized mid, Y is final

In [64]:
corr, _ = stats.pearsonr(normalized_mid, final)
print(corr)

0.544424742312


## (e) covariance between midterm-original and finals-original is 0.54442474

covariance = sum(xi-x_mean)(yj-y_mean)/N

In [66]:
np.corrcoef(mid,final)

array([[ 1.        ,  0.54442474],
       [ 0.54442474,  1.        ]])

# Question 4

In [76]:
import pandas as pd
lib = pd.read_table("CS_412_Fall21_Assignment1/data.libraries.inventories.txt")
lib

library  book_1  book_2  book_3  book_4  book_5  book_6  book_7  book_8  \
0     CML      53     136      90     103      36      93     195      51   
1     CBL     103      80     132     167     144     172     108      89   

   book_9    ...     book_91  book_92  book_93  book_94  book_95  book_96  \
0     192    ...          50      144       46      110      163      183   
1     103    ...          92      139      122      123      123       61   

   book_97  book_98  book_99  book_100  
0      107      129      162       173  
1      135       90      125       118  

[2 rows x 101 columns]

## (a) Minkowski distance of the vectors for CML and CBL with regard to different h values:

In [102]:
CML = lib.values[0][1:101]
CML

array([53, 136, 90, 103, 36, 93, 195, 51, 192, 148, 79, 30, 48, 80, 242,
       77, 170, 137, 41, 146, 116, 186, 101, 173, 110, 50, 149, 139, 30,
       119, 86, 97, 25, 63, 173, 130, 60, 148, 129, 69, 167, 83, 121, 28,
       133, 173, 149, 105, 64, 134, 137, 21, 179, 116, 35, 119, 60, 140,
       150, 68, 145, 197, 80, 73, 128, 3, 180, 96, 111, 95, 122, 132, 64,
       181, 165, 129, 79, 63, 72, 187, 99, 143, 49, 131, 84, 179, 215, 112,
       76, 64, 50, 144, 46, 110, 163, 183, 107, 129, 162, 173], dtype=object)

In [103]:
CBL = lib.values[1][1:101]
CBL

array([103, 80, 132, 167, 144, 172, 108, 89, 103, 63, 65, 58, 141, 116,
       152, 122, 129, 83, 175, 116, 173, 67, 61, 114, 195, 166, 85, 42, 87,
       183, 217, 125, 109, 71, 93, 71, 77, 209, 134, 59, 99, 147, 75, 50,
       181, 102, 173, 160, 147, 97, 90, 57, 151, 100, 102, 181, 118, 163,
       141, 146, 149, 149, 250, 128, 155, 148, 123, 189, 80, 186, 80, 114,
       102, 31, 66, 34, 59, 213, 104, 251, 31, 75, 164, 214, 53, 79, 118,
       75, 169, 192, 92, 139, 122, 123, 123, 61, 135, 90, 125, 118], dtype=object)

In [104]:
from scipy.spatial import distance

## h = 1: distance = 6152

In [105]:
distance.minkowski(CML, CBL, 1)

6152

## h = 2: distance = 715.3278968417211

In [106]:
distance.minkowski(CML, CBL, 2)

715.3278968417211

## h = infinity: distance = 170

###### as p approaches infinity, the Minkowski distance approaches maximum difference of coordinates. We obtain the Chebyshev distance.

In [150]:
from scipy.spatial import distance
distance.chebyshev(CML,CBL)

170

## (b) The cosine similarity between the feature vectors for CML and CBL is 0.84140402566230776

Cos(x, y) = x . y / ||x|| * ||y||

In [114]:
from scipy import spatial

result = 1 - spatial.distance.cosine(CML, CBL)
result

0.84140402566230776

spatial.distance.cosine computes the distance, and not the similarity. So we subtract the value from 1 to get the similarity.

## (c) the Kullback-Leibler (KL) divergence between CML and CBL is D(CML||CBL) = 0.20708093733159472

First we build distribution for both CML and CBL Ki = Ii/(I1+I2+...+I100)

In [145]:
CML_distribution = CML/sum(CML)
CML_distribution = CML_distribution.astype('float64')
sum(CML_distribution)

1.0000000000000002

In [146]:
CBL_distribution = CBL/sum(CBL)
CBL_distribution = CBL_distribution.astype('float64')
sum(CBL_distribution)

1.0000000000000002

###### calculate the kl divergence<br>
def kl_divergence(p, q):<br>
>return sum(p[i] * log2(p[i]/q[i]) for i in range(len(p)))

In [147]:
# we use scipy function to solve the answer.
from scipy.stats import entropy
entropy(CML_distribution,CBL_distribution)

0.20708093733159472